### PR TITLE
refactor(core): Abstract away duplication when finding first pinned trigger (no-changelog)

### PR DESCRIPTION
**Pull request by michael-radency**
Github issue / Community forum post (link here to close automatically):

### CODE REVIEW

File: packages/cli/src/WorkflowHelpers.ts [hunk 1]
```diff
@@ -566,3 +566,19 @@ export function validateWorkflowCredentialUsage(
 
 	return newWorkflowVersion;
 }
+
+//get starting node, overwise return undefined
```
Comment (ivov) [inline packages/cli/src/WorkflowHelpers.ts:570]: Prolly no need for this comment.

File: packages/cli/src/WorkflowHelpers.ts [hunk 2]
```diff
@@ -566,3 +566,19 @@ export function validateWorkflowCredentialUsage(
...
+export function getStartNode(
+	data: IWorkflowExecutionDataProcess,
+	workflow: Workflow,
+): INode | undefined {
```
Comment (ivov) [inline packages/cli/src/WorkflowHelpers.ts:574]: Return types are inferrable:

```suggestion
) {
```

**Review by ivov — COMMENTED** [review]
1. We have these three, a bit confusing. These cannot be combined?
- `getStartNode` at `packages/cli/src/WorkflowHelpers.ts`
- `__getStartNode` at `packages/workflow/src/Workflow.ts`
- `getStartNode` at `packages/workflow/src/Workflow.ts`
2. Since this is important functionality that it is now a separate function, we could consider unit-testing it.

**Review by edumaciel10 — COMMENTED** [review]
Looks good

File: packages/cli/src/WorkflowHelpers.ts [hunk 3]
```diff
@@ -566,3 +566,19 @@ export function validateWorkflowCredentialUsage(
...
+	let startNode;
+	if (
+		data.startNodes?.length === 1 &&
+		Object.keys(data.pinData ?? {}).includes(data.startNodes[0])
+	) {
+		startNode = workflow.getNode(data.startNodes[0]) ?? undefined;
+	}
+
+	return startNode;
```
[the remark this answers is not in the document]
Comment (michael-radency) [reply packages/cli/src/WorkflowHelpers.ts:575]: I leave return as it is now, because ts complains about not all return paths covered

**Comment by michael-radency** [comment]
> 1. We have these three, a bit confusing. These cannot be combined?
> 
> * `getStartNode` at `packages/cli/src/WorkflowHelpers.ts`
> * `__getStartNode` at `packages/workflow/src/Workflow.ts` (private used in getStartNode)
> * `getStartNode` at `packages/workflow/src/Workflow.ts`

I renamed added helper to getExecutionStartNode, they are not doing the same as far as I can tell

**Review by ivov — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.3.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.3.0)

**Pull request merged**